### DISCUSSION TITLE
Blank type annotations

### DISCUSSION THREAD

**Discussion by hkolbeck** [kdl-org/kdl#177] [General]
I'm starting to update [kdl4j](https://github.com/hkolbeck/kdl4j), and I ran across an edge case I thought I'd check feelings on. Right now you can specify a blank type annotation with `("")value`. I don't see any reason implementations should deny that, but it's worth keeping an eye on for round tripping KDL.

**Comment by zkat** [comment]
I'm fine with this, tbh

**Comment by Patitotective** [comment]
I think blank type annotations are pointless, they just adds more complexity to the parser having to make represent them as `Option` types (since empty does not really mean no value).

**Reply by hkolbeck** [reply]
That's a good point. I'd propose having a blank string type annotation be explicitly equivalent to no annotation, with a test whose output has no annotation.

**Reply by Patitotective** [reply]
Ok, then, what would the way to propose this? a PR changing the tests?
Thanks for the answer.

**Reply by zkat** [reply]
yes

**Comment by larsgw** [comment]
I also think it's very unnatural for the KDL document

    node

to result in

    { name: "node", tag: "" }

I think making `node` and `("")node` equivalent adds unnecessary complexity to the language itself. I would be fine with disallowing zero-length identifiers altogether, but that doesn't solve the apparent complexity with dealing with `null`/`Option` (though I can't really think of a situation where that would be a big problem).

**Reply by Patitotective** [reply]
I don't get why it is unnatural to get no type annotation when you specified no type annotation.

**Reply by larsgw** [reply]
I think it's unnatural to represent an absence with `""` when `null`, `undefined`, and even omitting the key entirely is also available in this JS example. The same goes for `Option` for Rust.

**Reply by Patitotective** [reply]
Ok, got the point.
Thanks.

**Reply by Patitotective** [reply]
I'm making a parser for the Nim programming language and since I'm going to make annotations `Option` types, should I also make children `Option` types?

**Reply by zkat** [reply]
That one's up to you, but [according to the proposed data model](https://github.com/kdl-org/kdl/pull/228/files#diff-6e91da909981c74d69a9fba6296f88490aa396e37c48f826c443edec4cb95935R73-R74), which hasn't been merged yet, they should be treated as semantically equivalent. But this data model doesn't need to correspond to your concrete implementation, it's just a mental model of what KDL is supposed to mean.

So, the answer is it'll depend on the needs of your parser: for example, [kdl-rs makes it an `Option<KdlNode>` type](https://github.com/kdl-org/kdl-rs/blob/main/src/node.rs#L23) because it's a formatting-preserving parser and whether or not to have those `{}` actually matters.

**Reply by tabatkins** [reply]
And conversely, [kdlpy](https://github.com/tabatkins/kdlpy) does *not* make such a distinction - the children are just exposed as a list, and both an empty `{}` and no braces at all result in an empty list. It's not a formatting-preserving parser, however, and the distinction between the two cases doesn't seem meaningful in most cases. (Whereas an explicitly-specified `("")` might actually be meaningfully distinct from no tag at all, so kdlpy uses `str | None` as the type for tags.)

**Reply by larsgw** [reply]
I agree. I'd argue that the tag equivalent to `{}` would be `()node`, not `("")node`. And that's already not valid (which makes sense to me). I guess `node` and `node {}` are a *bit* like `<div></div>` and `<div />`.

**Reply by tabatkins** [reply]
> a bit like `<div></div>` and `<div />`.

Which, notably, are exactly equivalent in XML (they're just two syntax forms of the exact same DOM, and can be used interchangeably). ^_^
